pkg/controller/unifiedpushserver/unifiedpushserver_controller.go

```diff
@@ -100,51 +103,184 @@ func (r *ReconcileUnifiedPushServer) Reconcile(request reconcile.Request) (recon
+			"database-name":     "unifiedpush",
+			"database-user":     "unifiedpush",
+			"database-password": "whateverups",
+		},
+	}
+}
+
+func newDeploymentForCR(cr *aerogearv1alpha1.UnifiedPushServer) *appsv1.Deployment {
 	labels := map[string]string{
-		"app": cr.Name,
+		"app":     cr.Name,
+		"service": "ups",
 	}
-	return &corev1.Pod{
+	return &appsv1.Deployment{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      cr.Name + "-pod",
+			Name:      cr.Name,
 			Namespace: cr.Namespace,
 			Labels:    labels,
 		},
-		Spec: corev1.PodSpec{
-			Containers: []corev1.Container{
-				{
-					Name:    "ups",
-					Image:   cr.Spec.Image,
-					Command: []string{"sleep", "3600"},
+		Spec: appsv1.DeploymentSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: labels,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:            "ups",
+							Image:           cr.Spec.Image,
+							ImagePullPolicy: corev1.PullAlways,
+							Env: []corev1.EnvVar{
+								{
+									Name:  "POSTGRES_SERVICE_HOST",
+									Value: "ups-postgres",
+								},
+								{
+									Name:  "POSTGRES_SERVICE_PORT",
+									Value: "5432",
+								},
+								{
+									Name: "POSTGRES_USER",
+									ValueFrom: &corev1.EnvVarSource{
+										SecretKeyRef: &corev1.SecretKeySelector{
+											Key: "database-user",
+											LocalObjectReference: corev1.LocalObjectReference{
+												Name: cr.Name,
+											},
+										},
+									},
+								},
+								{
+									Name: "POSTGRES_PASSWORD",
+									ValueFrom: &corev1.EnvVarSource{
+										SecretKeyRef: &corev1.SecretKeySelector{
+											Key: "database-password",
+											LocalObjectReference: corev1.LocalObjectReference{
+												Name: cr.Name,
+											},
+										},
+									},
+								},
+								{
+									Name: "POSTGRES_DATABASE",
+									ValueFrom: &corev1.EnvVarSource{
+										SecretKeyRef: &corev1.SecretKeySelector{
+											Key: "database-name",
+											LocalObjectReference: corev1.LocalObjectReference{
+												Name: cr.Name,
+											},
+										},
+									},
+								},
+							},
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "ups",
+									Protocol:      corev1.ProtocolTCP,
+									ContainerPort: 8080,
+								},
+							},
+							ReadinessProbe: &corev1.Probe{
+								Handler: corev1.Handler{
+									HTTPGet: &corev1.HTTPGetAction{
+										Path: "/rest/applications",
+										Port: intstr.IntOrString{
+											Type:   intstr.Int,
+											IntVal: 8080,
+										},
+									},
+								},
+								InitialDelaySeconds: 15,
+								TimeoutSeconds:      2,
+							},
+						},
+						{
+							Name:            "ups-oauth-proxy",
+							Image:           "docker.io/openshift/oauth-proxy:v1.1.0", // TODO: allow overriding this
```

Comment:
Should this also be configurable through the CR like the UPS one? Or is that the right place for defining images at all?

If they should be in the CR, should it be like:

```yaml 
# [...]
spec:
  upsImage: docker.io/aerogear/unifiedpush-wildfly-plain:2.2.1.Final
  oauthProxyImage: docker.io/openshift/oauth-proxy:v1.1.0
# [...]
```
or more like:
```yaml 
# [...]
spec:
  images:
    unifiedpush: docker.io/aerogear/unifiedpush-wildfly-plain:2.2.1.Final
    oauthproxy: docker.io/openshift/oauth-proxy:v1.1.0
# [...]
```

or something else?

Reply:
I like the second option. I think these values should be used as default if none is provided. Use the values from the CR file if they are defined. This will allow us to override them to use the productised images if required.

Reply:
Actually, think more about it, I think these values should be set via env vars, rather than in the CR. The reason is that these values are not something that can be changed by the CRs: what happens when there are 2 CRs specify different values for the images? One of them surely will not work. But if we use env var, it won't be a problem.

Reply:
Yeah, that's a good idea I think. Where would we specify those env vars though? In the Dockerfile for the operator build?

Reply:
I think it is ok to set the default values in the code, and only checks the env var when it is set. If nothing is set just use the default one. This way we don't need to define them in dockerfiles, but can still override them in other environments.